packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp

```diff
@@ -0,0 +1,404 @@
+        merged.push_back(indices1[ix[i1++]]);
+    }
+    std::copy(indices2+i2, indices2+n2, std::back_inserter(merged));
+  }
+
+  // Return only the unique elements
+  auto last = std::unique(merged.begin(), merged.end());
+  merged.erase(last, merged.end());
+  return merged;
+}
+
+/// \brief Copies the elements from the array \c indices1 which are
+///   not found in the array \c indices2 to the array \c diff.
+///
+///   On input, indices1 need not be sorted, but indices2 must be sorted. The
+///   intended use case is indices1 contains new elements to merge in to the
+///   existing CRS indices contained in indices2 - which should already be
+///   sorted.
+///
+///   This function is modeled after std::set_difference but takes arrays and
+///   lengths instead of iterators. It also does not require that both input
+///   arrays be sorted (just the second).
+///
+/// \param [in] indices1 - the first array of indices
+/// \param [in] n1 - the length of indices1
+/// \param [in] indices2 - the second array of indices
+/// \param [in] n2 - the length of indices2
+///
+/// \return diff - The difference between \c indices1 and \c indices2
+///
+template <class Ordinal>
+Teuchos::Array<Ordinal>
+ind_difference(
+    Ordinal const * const indices1,
+    size_t const n1,
+    Ordinal const * const indices2,
+    size_t const n2)
+{
+  Teuchos::Array<Ordinal> diff;
+
+  // We assume that indices1 is not sorted. It is the index array sent to us by
+  // users. indices2, on the other hand, should be sorted. It is the existing
+  // CRS index array.
+  auto ix = argsort(indices1, n1);
+
+  size_t i1 = 0;
+  size_t i2 = 0;
+  while (i1 < n1)
+  {
+    if (i2 == n2)
+    {
+      for (size_t i = i1; i < n1; i++)
+        diff.push_back(indices1[ix[i]]);
+      break;
+    }
+    auto ind1 = indices1[ix[i1]];
+    auto ind2 = indices2[i2];
+    if (ind1 < ind2)
+    {
+      diff.push_back(ind1);
+      i1++;
+    }
+    else
+    {
+      if (! (ind2 < ind1))
+        i1++;
+      i2++;
+    }
+  }
+
+  // Return only unique elements
+  auto last = std::unique(diff.begin(), diff.end());
+  diff.erase(last, diff.end());
+  return diff;
+}
+
+/// \brief Implementation of insertCrsIndices
+template<class T>
+int
+insert_crs_indices(
+    T& indices,
+    size_t const num_assigned,
+    typename T::value_type const in_indices[],
+    size_t const num_in)
+{
+  if (num_in == 0)
+    return 0;
+
+  // Rather than loop through the two arrays of indices and looking for their
+  // difference and then checking if their is enough capacity in indices to
+  // accommodate the new elements (not already in indices, excluding
+  // duplicates), we go ahead and merge the two arrays in to one. After the
+  // arrays are merged, we can check if the length of the merged array exceeds
+  // the capacity of the indices array. If it does, return -1 to indicate the
+  // error. If it does not, however, we can just copy the merged elements back
+  // in to \c indices.
```

Comment:
OK, I think I see the value here of allocating a new array to return from `ind_merge` -- the idea is that you want to know whether there is enough space before destructively modifying the graph.  I'm OK with that.

Reply:
Right.  To use any of the `std` sort or merge procedures we’d have to allocate a sorted copy of the user’s input array, so why not just create a new super merged array - which allows us to leave the original graph unmodified if there is not capacity for the new stuff, eh?